Ocaml enhancements

ericbmerritt opened the pull request:
This adds several ocaml packages to the system, and fixes/refactors a couple. In general the commits go in dependency order. 

I added a commit to upgrade the default ocaml to 4.02.1. Its been out quite a while and is stable. However, this is potentially quite a big change and may have unintended consequences.  So at the maintainers request I will drop it. 

On OSX this requires #7732 to build.

ericbmerritt committed:
add ericbmerritt to the maintainers list

ericbmerritt committed:
ocaml-typeconv: upgrade to 112.01.01

ericbmerritt committed:
ocaml-bin_prot: add initial version (112.24.00) to the system

ericbmerritt committed:
ocaml-custom_printf: add initial version (112.24.00) to the system

ericbmerritt committed:
ocaml-comparelib: add initial version (109.60.00) to the system

ericbmerritt commented:
the build failure seems to be due to transient connectivity problems on the travis hosts.

vbgl commented:
Nice. I’m going to give a look to these commits.

About making ocaml-4.02 the default, I think we need to fix the camlp4 issue first. See http://caml.inria.fr/mantis/view.php?id=6605 for details. In particular, `ulex` and `optcomp` are broken.

Why do you use `dontStrip = true` everywhere? (I’m not saying it’s an error; I’d just like to understand).

`ctypes` does not build on my system (chrooted fedora linux_x64). It fails at finding `libffi`. Might be an upstream issue with their `configure` script. Quick and dirty workaround to be performed after `make libffi.config`:

```
    substituteInPlace libffi.config \
      --replace 'libffi_available=false' 'libffi_available=true'
```

ericbmerritt commented:
@vbgl I will drop the 4.02.1 update patch then.  On the dontString, its there just because the stripper tries to strip the ocaml binaries, cma, cmxa etc. It doesn't recognize the format so you get a bunch of warnings about those files being unrecognized. Its not a big deal, but just seems dirty. The dontStrip solves it, but there may be some unintended consequences there. 

I will make the rest of your suggested changes and update the PR.

vbgl commented:
About `ctypes`: the `libffi_config` issue can be solved by adding `ncurses` to the `buildInputs`. The `discover` script finds `libffi` using `pkg-config`. It then builds a test; this fails because the `curses` library cannot be found.

ericbmerritt commented:
@vbgl this is updated in response to your comments. Be aware though that several of these packages require 4.02.1 so you will have to over ride and set ocaml to 4.02.1 to test these. I also added re2, textutils, and core_extended. I usually don't like adding to a pr once its in flight, but given the nature of this one it made sense.

ericbmerritt committed:
ocaml-herelib: add initial version (109.35.02) to the system

ericbmerritt committed:
ocaml-pa_test: add initial version (112.24.00) to the system

ericbmerritt committed:
ocaml-pipebang: add initial version (110.01.00) to the system

ericbmerritt committed:
ocaml-typerep: add initial version (112.24.00) to the system

ericbmerritt committed:
ocaml-async_kernel: add initial version (112.24.00) to the system

ericbmerritt committed:
ocaml-async: add initial version (112.24.00) to the system

ericbmerritt committed:
ocaml-ctypes: add initial version (0.4.1) to the system

ericbmerritt committed:
ocaml-cstruct: fix various build problems

ericbmerritt committed:
ocaml-ipaddr: add version (2.6.1) to the system

ericbmerritt committed:
ocaml-conduit: add initial version (0.8.3) to the system

ericbmerritt committed:
ocaml-alcotest: add initial version (0.3.1) to the system

ericbmerritt committed:
ocaml-cohttp: add initial version (0.17.1) to the system

ericbmerritt committed:
ocaml-re2: add initial version (112.06.00) to the system

ericbmerritt committed:
ocaml-jingoo: add initial version (1.2.7) to the system

ericbmerritt commented:
@vbgl this should address your existing comments. A few things to note. I added one more package (jingoo) als renamed ocaml_sexplib to sexplib (your herelib comment struck home and the whole sexplib thing bothered me). Also I noticed that I had duplicated ipaddr, I missed that there was an existing package named ocaml_ipaddr. I merged those in the ipaddr commit.

vbgl commented:
`jingoo` fails during installation. You need something like that:

```
preInstall = "mkdir -p $out/bin";
installFlags = "BINDIR=$(out)/bin";
```

ericbmerritt commented:
@vbgl thanks, I will get that fixed. I really wonder why this isn't failing for me though.

Fuuzetsu commented:
The stripper warnings are just that, warnings, and we shouldn't be putting `dontStrip = true` everywhere to skip them. Maybe in the future we will happen to need to strip something out and we'll be none the wiser.

If in general we don't want to strip for ocaml stuff, it might be worthwhile having a small builder for ocaml that sets dontStrip as well as automatically supplies ocaml as buildInput and other such stuff.

ericbmerritt commented:
I am in strong disagreement with that statement. Warnings should be taken seriously or removed, otherwise they loose their value. That said, I am not running this project so I will remove them so this can get merged.

Fuuzetsu commented:
Sure but your fix is nothing better than just removing the thing causing the warnings all together which isn't exactly proper either.

Anyway it is not so serious that your PR wouldn't get merged, just a comment.

vbgl commented on pkgs/development/ocaml-modules/cohttp/0.17.1.nix:
```diff
@@ -0,0 +1,36 @@
+
+  buildInputs = [ocaml findlib alcotest];
+  propagatedBuildInputs = [camlp4 cmdliner re uri fieldslib sexplib sexplib
+                           conduit stringext base64 magic-mime ounit async
```
There is one extra `sexplib` trailing. In fact no occurrence is needed at all, since it’s propagated by, e.g., `conduit`.

ericbmerritt commented:
@Fuuzetsu Well, the warnings are spurious in this case. The strip command can't strip ocaml code, so it shouldn't be run. The 'dontStrip' solves the root cause of the problem (strip shouldn't be run on ocaml code). However, it would be nice if that was just automatic. I think your suggestion of a small derivation for ocaml is probably the right way to go. I just need to figure that out, I am still a very new to nix.

@vbgl Thanks for the comments. I will get these resolved shortly. I am traveling today and tomorrow so it will probably be friday before I can do anything.

Fuuzetsu commented:
@ericbmerritt for examples of some existing builders peek in `pkgs/build-support`. Grepping there for `inherit` could be instructive.

ericbmerritt committed:
add an ocaml specific builder to the system

ericbmerritt commented:
@Fuuzetsu Please take a look at the 3cf8728. If this looks like a reasonable approach I will refactor the existing commits to use it. If its not reasonable, I would love direction to a more reasonable approach.

Fuuzetsu commented:
@ericbmerritt Looks reasonable to me though mind I don't use ocaml. If it works for you then great.

I'd suggest putting a small comment at the top of the builder about who to contact about it (you). Sure there's git history but it can be pretty tedious to try and dig through it if many people touch the file through various directory structure changes and so on.

ericbmerritt commented:
I will do it and refactor my current commits to use it. Thanks!

ericbmerritt committed:
ocaml-pa_ounit: add initial version (112.24.00) to the system

ericbmerritt committed:
ocaml-pa_bench: add initial version (112.06.00) to the system

ericbmerritt committed:
ocaml-lwt: make sure libev propagates as a dependency

ericbmerritt commented:
@Fuuzetsu @vbgl this latest set of changes covers all of your comments, including the derivation for @Fuuzetsu and the re2 change for @vbgl along with quite a bit of cleanup.

ericbmerritt committed:
ocaml-opam: the sha hash of 1.2.2 changed

This is a bit scary. The sha of 1.2.2 changed, causing a crash on
download of the url. This updates to the current sha. Opam maintainer
why, oh why, do you change a released version without a version bump??

ericbmerritt committed:
sexplib: rename from ocaml_sexplib to sexplib

There really isn't any need for it to be named ocaml_sexplib

ericbmerritt committed:
ocaml-core: add initial version (112.24.01) to the system

ericbmerritt committed:
rename ocaml_typeconv to type_conv

The actual name of the package is type_conv
(https://github.com/janestreet/type_conv). Having it named
ocaml_typeconv adds a verbosity that isn't required and actually makes
the package harder to find.

ericbmerritt committed:
ocaml-sexplib: upgrade to 112.24.01

ericbmerritt committed:
ocaml-textutils: add initial version (112.17.00) to the system

ericbmerritt committed:
ocaml-variantslib: add initial version (109.15.03) to the system

ericbmerritt committed:
ocaml-core_kernel: add initial version (112.24.00) to the system

ericbmerritt committed:
ocaml-async_unix: add initial version (112.24.00) to the system

ericbmerritt committed:
ocaml-async_extra: add initial version (112.24.00) to the system

ericbmerritt committed:
ocaml-async_ssl: add initial version (112.24.03) to the system

ericbmerritt committed:
ocaml-core_extended: add initial version (112.24.00) to the system

ericbmerritt committed:
ocaml-async_find: add initial version (111.28.00) to the system

ericbmerritt committed:
ocaml-async_shell: add initial version (109.28.03) to the system

ericbmerritt commented:
This latest set of refactoring moves the packages to use 'default.nix'. It also adds `hasSharedObjects = true;` in several packages that needed it.

vbgl commented:
`re2` fails to build (on chrooted linux/x86_64):

```
ld: cannot find -lc++
```

I suspect that the patches should only be applied on darwin/OSX.

ericbmerritt commented:
@vbgl yes. crap. I will fix that.

ericbmerritt commented:
@vbgl fixed and tested this time. Sorry about that.

ericbmerritt committed:
add jwilberding as a maintainer

ericbmerritt committed:
ocaml-atd: add initial version (1.1.2) to the system

ericbmerritt committed:
ocaml-atdgen: add minimal version (1.6.0) to the system

ericbmerritt commented:
Nice!! Thanks!@vbgl!
